File: problems/python/parking_lot_system/parking_lot.py
import datetime
import math

class Vehicle:
    def __init__(self, spot_size):
        self.spot_size = spot_size
    
    def get_spot_size(self):
        return self.spot_size

class Car(Vehicle):
    def __init__(self):
        super().__init__(1)

class Limo(Vehicle):
    def __init__(self):
        super().__init__(2)

class Truck(Vehicle):
    def __init__(self):
        super().__init__(3)

class Driver:
    def __init__(self, id, vehicle):
        self.id = id
        self.vehicle = vehicle
        self.payment_due = 0
    
    def get_id(self):
        return self.id

    def get_vehicle(self):
        return self.vehicle

    def charge(self, amount):
        self.payment_due += amount

class ParkingFloor:
    def __init__(self, number_of_spots):
        self.spots = [0] * number_of_spots
        self.vehicle_map = {}

    def park_vehicle(self, vehicle):
        vehicle_spot_size = vehicle.get_spot_size()
        l = r = 0

        while r < len(self.spots):
            if self.spots[r] != 0:
                l = r + 1
            # able to park this vehicle at sport [l, r]
            if r - l + 1 == vehicle_spot_size:
                for k in range(l, r + 1):
                    self.spots[k] = 1
                self.vehicle_map[vehicle] = [l, r]
                return True
            r += 1
        return False
    
    def remove_vehicle(self, vehicle):
        l, r = self.vehicle_map[vehicle]
        for k in range(l, r + 1):
            self.spots[k] = 0

        del self.vehicle_map[vehicle]
    
    def get_parking_spots(self):
        return self.spots

    def get_vehicle_spots(self, vehicle):
        return self.vehicle_map.get(vehicle, [])

class ParkingGarage:
    def __init__(self, number_of_floors, spots_per_floor):
        self.parking_floors = [ParkingFloor(spots_per_floor) for _ in range(number_of_floors)]

    def park_vehicle(self, vehicle):
        for floor in self.parking_floors:
            if floor.park_vehicle(vehicle):
                return True
        return False

    def remove_vehicle(self, vehicle):
        for floor in self.parking_floors:
            if floor.get_vehicle_spots(vehicle):
                floor.remove_vehicle(vehicle)
                return True
        return False

class ParkingSystem:
    def __init__(self, parking_garage, hourly_rate):
        self.parking_garage = parking_garage
        self.hourly_rate = hourly_rate
        self.checkedin_time = {} # map driver id: time

    def park_vehicle(self, driver):
        curr_hour = datetime.datetime.now().hour
        is_parked = self.parking_garage.park_vehicle(driver.get_vehicle())
        if is_parked:
            self.checkedin_time[driver.get_id()] = curr_hour
        return is_parked
    
    def remove_vehicle(self, driver):
        if driver.get_id() not in self.checkedin_time:
            return False
        curr_hour = datetime.datetime.now().hour
        checkedin_hour = self.checkedin_time[driver.get_id()]
        driver.charge((curr_hour - checkedin_hour) * self.hourly_rate)

        del self.checkedin_time[driver.get_id()]
        return self.parking_garage.remove_vehicle(driver.get_vehicle())

parkingGarage = ParkingGarage(2, 3)
parkingSystem = ParkingSystem(parkingGarage, 1)

driver1 = Driver(1, Car())
driver2 = Driver(2, Limo())
driver3 = Driver(3, Truck())
driver4 = Driver(4, Truck())

print(parkingSystem.park_vehicle(driver1))
print(parkingSystem.park_vehicle(driver2))
print(parkingSystem.park_vehicle(driver3))
print(parkingSystem.park_vehicle(driver4))

print(parkingSystem.remove_vehicle(driver1))
print(parkingSystem.remove_vehicle(driver4))

File: problems/cpp/parking_lot_system/parking_lot.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

class Vehicle {
private:
    int spotSize;

public:
    Vehicle(int spotSize) {
        this->spotSize = spotSize;
    }

    int getSpotSize() {
        return this->spotSize;
    }
};

class Car : public Vehicle {
public:
    Car(): Vehicle(1) {}
};

class Limo : public Vehicle {
public:
    Limo(): Vehicle(2) {}
};

class Truck : public Vehicle {
public:
    Truck(): Vehicle(3) {}
};

class Driver {
private:
    int id = 0;
    Vehicle* vehicle;
    int paymentDue;

public:
    Driver(int id, Vehicle* vehicle) {
        this->id = id;
        this->vehicle = vehicle;
        this->paymentDue = 0;
    }

    int getId() {
        return this->id;
    }

    Vehicle* getVehicle() {
        return this->vehicle;
    }

    void charge(int amount) {
        this->paymentDue += amount;
    }

};

class ParkingFloor {
private:
    vector<int> spots;
    unordered_map<Vehicle*, vector<int> > vehicleMap;

public:
    ParkingFloor(int numberOfSpots) {
        this->spots = vector<int>(numberOfSpots);
        this->vehicleMap = unordered_map<Vehicle*, vector<int> >();
    }

    bool parkVehicle(Vehicle* vehicle) {
        int vehicle_spot_size = vehicle->getSpotSize();
        int l = 0, r = 0;

        while (r < this->spots.size()) {
            if (this->spots[r] != 0) {
                l = r + 1;
            }
            // able to park this vehicle
            if (r - l + 1 == vehicle_spot_size) {
                for (int k = l; k <= r; k++) {
                    this->spots[k] = 1;
                }
                this->vehicleMap[vehicle] = vector<int> {l, r};
                return true;
            }
            r++;
        }
        return false;
    }

    bool removeVehicle(Vehicle* vehicle) {
        vector<int> curr_parked = this->vehicleMap[vehicle];
        for (int i = curr_parked[0]; i <= curr_parked[1]; i++) {
            this->spots[i] = 0;
        }
        this->vehicleMap.erase(vehicle);
    }

    vector<int> getParkingSpots() {
        return this->spots;
    }

    vector<int> getVehicleSpots(Vehicle* vehicle) {
        return this->vehicleMap[vehicle];
    }
};

class ParkingGarage {
private:
    vector<ParkingFloor*> parkingFloors;

public:
    ParkingGarage(int floorCounts, int spotsPerFloor) {
        this->parkingFloors = vector<ParkingFloor*>(floorCounts);
        for (int i = 0; i < floorCounts; i++) {
            this->parkingFloors[i] = new ParkingFloor(spotsPerFloor);
        }
    }

    bool parkVehicle(Vehicle* vehicle) {
        for (ParkingFloor* floor : this->parkingFloors) {
            if (floor->parkVehicle(vehicle)) {
                return true;
            }
        }
        return false;
    }

    bool removeVehicle(Vehicle* vehicle) {
        for (ParkingFloor* floor : this->parkingFloors) {
            if (floor->getVehicleSpots(vehicle).size() != 0) {
                floor->removeVehicle(vehicle);
                return true;
            }
        }
        return false;
    }
};

class ParkingSystem {
private:
    ParkingGarage* parkingGarage;
    int hourlyRate = 0;
    unordered_map<int, int> checkedInTime; // map driver id: checked in time

public:
    ParkingSystem(ParkingGarage* parkingGarage, int hourlyRate) {
        this->parkingGarage = parkingGarage;
        this->hourlyRate = hourlyRate;
        this->checkedInTime = unordered_map<int, int>();
    }

    bool parkVehicle(Driver* driver) {
        int currHour = time(0);
        bool isParked = this->parkingGarage->parkVehicle(driver->getVehicle());
        if (isParked) {
            this->checkedInTime[driver->getId()] = currHour;
            return true;
        }
        return false;
    }

    bool removeVehicle(Driver* driver) {
        if (this->checkedInTime.find(driver->getId()) == this->checkedInTime.end()) {
            cout << "Here" << endl;
            return false;
        }
        int currHour = time(0);
        int checkedInHour = this->checkedInTime[driver->getId()];
        driver->charge((currHour - checkedInHour) * this->hourlyRate);
        this->checkedInTime.erase(driver->getId());
        return this->parkingGarage->removeVehicle(driver->getVehicle());
    }
};

int main() {
    ParkingGarage* parkingGarage = new ParkingGarage(3, 3);
    ParkingSystem* parkingSystem = new ParkingSystem(parkingGarage, 1);

    Driver* driver1 = new Driver(1, new Car());
    Driver* driver2 = new Driver(2, new Limo());
    Driver* driver3 = new Driver(3, new Truck());
    Driver* driver4 = new Driver(4, new Truck());

    cout << parkingSystem->parkVehicle(driver1) << endl;
    cout << parkingSystem->parkVehicle(driver2) << endl;
    cout << parkingSystem->parkVehicle(driver3) << endl;
    cout << parkingSystem->parkVehicle(driver4) << endl;

    cout << parkingSystem->removeVehicle(driver1) << endl;
    cout << parkingSystem->removeVehicle(driver4) << endl;
    
}